chain/tron/builder.go

```diff
@@ -159,7 +170,352 @@ func (txBuilder TxBuilder) NewTokenTransfer(from xc.Address, to xc.Address, amou
+
+	contract := &core.UnfreezeBalanceV2Contract{}
+	contract.OwnerAddress = from_bytes
+	contract.UnfreezeBalance = balance.Int().Int64()
+	contract.Resource = core.ResourceCode_BANDWIDTH
+
+	params, err := ptypes.MarshalAny(contract)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal any params: %w", err)
+	}
+
+	tx_contract := &core.Transaction_Contract{
+		Type:      core.Transaction_Contract_UnfreezeBalanceV2Contract,
+		Parameter: params,
+	}
+
+	i := input.(*txinput.TxInput)
+	tx := new(core.Transaction)
+	tx.RawData = i.ToRawData(tx_contract)
+
+	return tx, nil
+}
+
+func (txBuilder TxBuilder) NewVotes(from xc.Address, votes []*httpclient.Vote, input *txinput.TxInput) (*core.Transaction, error) {
+	from_bytes, err := GetAddressHash(string(from))
+	if err != nil {
+		return nil, err
+	}
+
+	contract := &core.VoteWitnessContract{}
+	contract.OwnerAddress = from_bytes
+	contract.Votes = make([]*core.VoteWitnessContract_Vote, len(votes))
+
+	for i, v := range votes {
+		addrhash, err := GetAddressHash(string(v.VoteAddress))
+		if err != nil {
+			return nil, fmt.Errorf("failed to get super representative address hash: %w", err)
+		}
+		contract.Votes[i] = &core.VoteWitnessContract_Vote{
+			VoteAddress: addrhash,
+			VoteCount:   int64(v.VoteCount),
+		}
+	}
+
+	params, err := ptypes.MarshalAny(contract)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal any params: %w", err)
+	}
+
+	tx_contract := &core.Transaction_Contract{
+		Type:      core.Transaction_Contract_VoteWitnessContract,
+		Parameter: params,
+	}
+
+	tx := new(core.Transaction)
+	tx.RawData = input.ToRawData(tx_contract)
+
+	return tx, nil
+}
+
+func (txBuilder TxBuilder) Stake(stakingArgs xcbuilder.StakeArgs, input xc.StakeTxInput) (xc.Tx, error) {
+	stakingInput, ok := input.(*txinput.StakeInput)
+	if !ok {
+		return nil, errors.New("invalid input type")
+	}
+
+	validator, ok := stakingArgs.GetValidator()
+	if !ok {
+		return nil, buildererrors.ErrValidatorRequired
+	}
+
+	stakeAmount, ok := stakingArgs.GetAmount()
+	if !ok {
+		return nil, buildererrors.ErrStakingAmountRequired
+	}
+
+	stakeVotes := common.TrxToVotes(stakeAmount, stakingInput.Decimals)
+	err := validateStakeAmount(stakeAmount, stakeVotes, stakingInput.Decimals)
+	if err != nil {
+		return nil, err
+	}
+
+	var validatorVotes *httpclient.Vote
+	usedVotes := uint64(0)
+	for _, v := range stakingInput.Votes {
+		usedVotes += v.VoteCount
+		if v.VoteAddress == validator {
+			validatorVotes = v
+		}
+	}
+	freezedBalance := xc.NewAmountBlockchainFromUint64(stakingInput.FreezedBalance)
+	availableVotes := common.TrxToVotes(freezedBalance, stakingInput.Decimals)
+	unusedVotes := availableVotes - usedVotes
+
+	from := stakingArgs.GetFrom()
+	tx := NewTx()
+
+	// we have to freeze additional TRX amount to satisfy stake requirements
+	if unusedVotes < stakeVotes {
+		missingVotes := stakeVotes - unusedVotes
+		d := math.Pow10(stakingInput.Decimals)
+		rawAmt := uint64(float64(missingVotes) * d)
```

Comment:
Please use safe math methods by convention.   Floating point math may not even be deterministic on some platforms (understood chance of impact is small due to the truncation).